File: brokk-foreman/src/lib/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

// Channel names mirror src-tauri/src/events.rs.
export const ACP_UPDATE_CHANNEL = "acp:update";
export const ACP_SESSION_STATUS_CHANNEL = "acp:session_status";
export const ACP_PERMISSION_CHANNEL = "acp:permission";
export const ACP_STDERR_CHANNEL = "acp:stderr";

export interface AcpUpdate {
  session_id: string;
  update: unknown; // Raw SessionUpdate JSON from the agent.
}

export interface PermissionRequest {
  request_id: string;
  session_id: string;
  tool_call: unknown;
  options: unknown;
}

export type SessionStatus =
  | { kind: "starting"; agent_id: string }
  | { kind: "ready"; agent_id: string; session_id: string }
  | { kind: "auth_required"; agent_id: string; auth_kind: string }
  | { kind: "stopped"; agent_id: string; reason: string }
  | { kind: "failed"; agent_id: string; error: string };

export interface AcpStderr {
  agent_id: string;
  line: string;
}

export function onAcpUpdate(
  handler: (payload: AcpUpdate) => void,
): Promise<UnlistenFn> {
  return listen<AcpUpdate>(ACP_UPDATE_CHANNEL, (event) => handler(event.payload));
}

export function onSessionStatus(
  handler: (payload: SessionStatus) => void,
): Promise<UnlistenFn> {
  return listen<SessionStatus>(ACP_SESSION_STATUS_CHANNEL, (event) =>
    handler(event.payload),
  );
}

export function onPermissionRequest(
  handler: (payload: PermissionRequest) => void,
): Promise<UnlistenFn> {
  return listen<PermissionRequest>(ACP_PERMISSION_CHANNEL, (event) =>
    handler(event.payload),
  );
}

export function onStderr(
  handler: (payload: AcpStderr) => void,
): Promise<UnlistenFn> {
  return listen<AcpStderr>(ACP_STDERR_CHANNEL, (event) => handler(event.payload));
}

File: brokk-foreman/src-tauri/src/events.rs
//! Payload types for Tauri events emitted from the Rust backend to the
//! Svelte frontend. The channel names are stable wire identifiers.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Channel: every ACP `SessionUpdate` notification received from the active
/// agent. v1 has at most one live session so we don't suffix the channel
/// with a session id — the frontend mirrors the same constraint.
pub const ACP_UPDATE_CHANNEL: &str = "acp:update";

/// Channel: lifecycle transitions of the active session.
pub const ACP_SESSION_STATUS_CHANNEL: &str = "acp:session_status";

/// Channel: incoming `requestPermission` from the agent. The frontend
/// surfaces a modal and calls `respond_permission(req_id, accept)` back.
pub const ACP_PERMISSION_CHANNEL: &str = "acp:permission";

/// Channel: stderr lines from the spawned agent process (best-effort tail
/// for the agents page on failure).
pub const ACP_STDERR_CHANNEL: &str = "acp:stderr";

/// One forwarded `SessionUpdate`. We pass the raw schema notification as
/// JSON so the frontend can render new update variants without a Rust
/// release.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AcpUpdate {
    pub session_id: String,
    pub update: Value,
}

/// One forwarded `requestPermission`. The frontend keys its modal by
/// `request_id` and posts the user's choice back via the
/// `respond_permission` command.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionRequest {
    pub request_id: String,
    pub session_id: String,
    pub tool_call: Value,
    pub options: Value,
}

/// Lifecycle states the frontend cares about for the session pane.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum SessionStatus {
    Starting {
        agent_id: String,
    },
    Ready {
        agent_id: String,
        session_id: String,
    },
    AuthRequired {
        agent_id: String,
        auth_kind: String,
    },
    Stopped {
        agent_id: String,
        reason: String,
    },
    Failed {
        agent_id: String,
        error: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AcpStderr {
    pub agent_id: String,
    pub line: String,
}
